internal/controller/gitrepository_controller.go

```diff
@@ -1112,3 +1176,35 @@ func commitReference(obj *sourcev1.GitRepository, commit *git.Commit) string {
 	}
 	return commit.String()
 }
+
+// requiresVerification inspects a GitRepository's verification spec and its status
+// to determine whether the Git repository needs to be verified again. It does so by
+// first checking if the GitRepository has a verification spec. If it does, then
+// it returns true based on the following three conditions:
+//
+//   - If the object does not have a observed verification mode in its status.
+//   - If the observed verification mode indicates that only the tag had been
+//     verified earlier and the HEAD also needs to be verified now.
+//   - If the observed verification mode indicates that only the HEAD had been
+//     verified earlier and the tag also needs to be verified now.
+func requiresVerification(obj *sourcev1.GitRepository) bool {
+	var verificationRequired bool
+	if obj.Spec.Verification != nil {
+		observedVerificationMode := obj.Status.ObservedVerificationMode
+		mode := obj.Spec.Verification.GetMode()
+		if observedVerificationMode == nil {
+			verificationRequired = true
+		} else if *observedVerificationMode == sourcev1.ModeTag &&
+			(mode == sourcev1.ModeHEAD || mode == sourcev1.ModeTagAndHEAD) {
+			verificationRequired = true
+		} else if *observedVerificationMode == sourcev1.ModeHEAD &&
+			(mode == sourcev1.ModeTag || mode == sourcev1.ModeTagAndHEAD) {
+			verificationRequired = true
+		}
+	}
+	return verificationRequired
+}
```

Comment:
```suggestion
func requiresVerification(obj *sourcev1.GitRepository) bool {
	if obj.Spec.Verification != nil {
		observedMode := obj.Status.ObservedVerificationMode
		mode := obj.Spec.Verification.GetMode()

		if observedMode == nil {
			return true
		}

		if (*observedMode == sourcev1.ModeTag && (mode == sourcev1.ModeHEAD || mode == sourcev1.ModeTagAndHEAD)) ||
	 	  (*observedMode == sourcev1.ModeHEAD && (mode == sourcev1.ModeTag || mode == sourcev1.ModeTagAndHEAD)) {
			return true
		}
	}
	return false
}
```

might be a bit easier to read?